Libvirt SELinux

5umm3r15 opened the pull request:
New SELinux policy for libvirt drivers:
Hypervisor drivers:
 - virtqemud  (QEMU/KVM)
 - virtlxcd   (LXC)
 - virtvboxd  (VirtualBox)

Secondary drivers:
 - virtstoraged   (host storage mgmt)
 - virtnetworkd   (virtual network mgmt)
 - virtinterface  (network interface mgmt)
 - virtnodedevd   (physical device mgmt)
 - virtsecretd    (security credential mgmt)
 - virtnwfilterd  (ip[6]tables/ebtables mgmt)
 - virtproxyd     (proxy daemon)

SELinux policy for virtvxz and virtxend has not been created yet.

In the first commit are services separated from the original virt files that are not libvirt related and virt_supplementary policy module is created.

In the second commit is designed policy for libvirt drivers, including modification of old virt interfaces and also creating new interfaces.

Copr build of SELinux policy with new libvirt policy:
https://copr.fedorainfracloud.org/coprs/nknazeko/selinux-policy/

Libvirt repository with SELinux subpackage is available on github:
https://github.com/5umm3r15/libvirt-selinux/tree/selinux

Copr build of Libvirt with SELinux subpackage: https://copr.fedorainfracloud.org/coprs/nknazeko/libvirt-selinux/

SELinux policy was also created for Libvirt Testsuite and used for testing virt policy. 
The repository is available on github: https://github.com/5umm3r15/perl-Sys-Virt-TCK-selinux/tree/selinux
Copr build of Libvirt-TCK with SELinux subpackage: https://copr.fedorainfracloud.org/coprs/nknazeko/libvirt-tck-selinux/

Libvirt test suite gives AVC messages related to this bug: 
[BZ1858260 - SELinux prevents svirt_t read|write on lvm_control_t during VM creation](https://bugzilla.redhat.com/show_bug.cgi?id=1858260)

5umm3r15 commented:
@zpytela @vmojzis @wrabcak Can you please review it?

wrabcak commented:
Thank you @5umm3r15 , what I don't understand is why is this PR against selinux-policy ?

5umm3r15 commented:
To be more transparent with changes against old policy.

5umm3r15 commented:
Hi! This is the final version of virt and virt_supplementary policies, which will be part of our policy.
Virt.* policy files will be also included in SELinux policy subpackage for Libvirt component.
@vmojzis @zpytela @wrabcak can you please review it? 
Thank you

WOnder93 commented on policy/modules/contrib/virt.te:
```diff
@@ -14,13 +14,19 @@ attribute virsh_transition_domain;
 attribute sandbox_caps_domain;
 
+permissive virtxend_t;
+permissive virtvzd_t;
```
There could also be a code comment explaining why these are permissive for now.

wrabcak replied:
There is also a possibility to add these domains for Fedora here: https://src.fedoraproject.org/rpms/selinux-policy/blob/master/f/permissivedomains.cil 

and we can keep policy sources without permissive domain.

WOnder93 replied:
I don't think that would make much sense. If the policy is absent, it simply wouldn't work without the `permissive`s. And it's not good to have this information in a different repo - it's *here* that the policy needs to be added in the future, so the permissive declarations should also be here. Then they can also be removed atomically in the same commit that adds the missing policy.

wrabcak replied:
No problem, as you say. :) In past, we kept permissive domains there. Also, if there is no SELinux policy for these types, what about unconfined_domain() macro? Using this macro we avoid lots of SELinux policy denials and confused users. WDYT?

WOnder93 replied:
That might indeed be a better approach, but I don't know enough about either approach to judge :)

zpytela replied:
I think the main question is: do we want to have users receive AVCs and report them or not?

IMHO in this stage of development should rather be permissive.

5umm3r15 replied:
I would prefer to have AVC's reported from Libvirt team, not users. For each driver I had many AVC messages and it might be confusing for users. Maybe it will be better, to use unconfined_domain macro until the AVC messages are reproduced internally and policy for these drivers is made.

vmojzis replied:
Agreed. I'd use the unconfined_domain here and then permissive domains in the policy libvirt-selinux will ship (this version will be tested by libvirt team). This way users can start using the new policy (split-daemon configuration will be used by default soon), we have time to polish the policy based on libvirt testing, and any issues still present in the final policy can be reported by users.

WOnder93 reviewed (commented):
I took a brief look at the changes and didn't find anything that would stick out as a problem. It looks very nice, good job!

Apart from the minor comment about documenting the permissive domains, I have one more style request/suggestion: Please try to use consistent whitespace. When modifying existing code, try to match the style of the surrounding code; when adding new code (new files) I'd suggest to use tabs, since those are used more often in the codebase:
```
$ git grep '    ' policy/modules/*/*.{te,if} | wc -l
6028
$ git grep '\t' policy/modules/*/*.{te,if} | wc -l
102854
```
The current state of whitespace is quite bad, but at least we could try to make it not worse / slightly better with new changes, IMHO :)

@zpytela What's your opinion on whitespace consistency?

zpytela commented:
Whitespace consistency is really poor in the whole policy context; I also prefer tabs to spaces in new policy, and only use spaces when the rest of the file uses them.

zpytela commented:
@5umm3r15 I see a lot of work has been done here, thank you for that. In this context I only have some minor comments.

1. Please convert all leading spaces to the right amount of tabs. Similarly do so in the comments inside an interface.
2. Block with booleans should be the first one in the te file.
3. Interfaces calls after kernel_* should be in alphabetical order in the te file.

I would also like to ask: Is this split meant temporarily or for good? It may be a bit confusing where to look for any virt_ interface or type in the future. In that case, we can at least have some brief explanation/description.

5umm3r15 committed:
Split virt policy, introduce virt_supplementary module

Separate the services from the original virt files that are not libvirt related and create virt_supplementary policy module.

5umm3r15 commented:
@zpytela @WOnder93 @wrabcak 
Thank you for the review. I replaced spaces by tabs, alphabetically ordered interfaces in kernel layer modules,
then in system layer modules, boolean blocks and optional policy blocks according to [style guide](https://github.com/TresysTechnology/refpolicy/wiki/StyleGuide).

The main virt policy, types and interfaces, is in the virt module, that will be handled by Libvirt. A lot of modules are using virt interfaces, so both virt and virt_supplementary modules will be in our policy for now. I am not sure if we decide to handle this problem another way in the future.

5umm3r15 committed:
Introduce SELinux policy for libvirt drivers

New SELinux policy for libvirt drivers:
Hypervisor drivers:
 - virtqemud  (QEMU/KVM)
 - virtlxcd   (LXC)
 - virtvboxd  (VirtualBox)

Secondary drivers:
 - virtstoraged   (host storage mgmt)
 - virtnetworkd   (virtual network mgmt)
 - virtinterface  (network interface mgmt)
 - virtnodedevd   (physical device mgmt)
 - virtsecretd    (security credential mgmt)
 - virtnwfilterd  (ip[6]tables/ebtables mgmt)
 - virtproxyd     (proxy daemon)

SELinux policy for virtvxz and virtxend has not been created yet and
these drivers run in unconfined domain.

Signed-off-by: Nikola Knazekova <nknazeko@redhat.com>

5umm3r15 committed:
MLS update for libvirt-selinux

Fix AVC messages discovered by testing on MLS system.

Move qemu_exec_t, qemu-storage-daemon and /var/log/log/* to virt_supplementary.

5umm3r15 committed:
Update policy for libvirt drivers

Allow libvirt log management daemon to search VM images dir.
Allow libvirt storage pool management daemon to manage VM images dir.
Allow virt drivers to create virt common files.

5umm3r15 committed:
Update policy for libvirt

Allow svirt to connect virtlogd socket.
Add net_admin capability to virtqemud, virtnetworkd and virtnodedevd.
Allow virtnetworkd to create dnsmasq_var_run_t dir.
Allow virtnetworkd to dbus chat with firewalld.

Allow virtqemud, virtnodedevd and virtstoraged to set priority of process.
Allow virtqemud to rw netlink_audit_socket.
Allow virtqemud to dbus chat with policykit and systemd machined
Allow virtqemud to delete /dev/urandom char file
Allow virtqemud to rw lvm_control

5umm3r15 committed:
Add macro dev_delete_urand

Allow domain to delete char_files in the dev/urandom directory.

vmojzis committed:
Enable virt_supplementary module

vmojzis commented:
Rebased and added virt_supplementary to policy/modules.conf

vmojzis committed:
Virt: Fix issues reported by selint

virt.te:  807: (C): Permissions in av rule not ordered (read_lnk_file_perms before ioctl) (C-005)
virt.te: 1843: (C): Permissions in av rule not ordered (create before connect) (C-005)
virt.te: 2038: (C): Permissions in av rule not ordered (setuid before setgid) (C-005)
virt.if:   51: (C): No comment before interface definition for virt_stub_svirt_sandbox_file (C-004)
virt.if:   82: (W): Attribute virt_image_type is listed in require block but not used in interface (W-003)
virt.if:   83: (W): Attribute virt_tmpfs_type is listed in require block but not used in interface (W-003)
virt.if:  100: (W): Type qemu_exec_t is used in interface but not required (W-002)
virt.if:  151: (W): Type virt_common_var_run_t is used in interface but not required (W-002)
virt.if:  152: (W): Type virt_common_var_run_t is used in interface but not required (W-002)
virt.if:  153: (W): Type virt_common_var_run_t is used in interface but not required (W-002)
virt.if:  154: (W): Type virt_common_var_run_t is used in interface but not required (W-002)
virt.if:  155: (W): Type virt_common_var_run_t is used in interface but not required (W-002)
virt.if:  876: (W): Type virt_var_lib_t is listed in require block but not used in interface (W-003)
virt.if:  971: (W): Type virt_var_lib_t is listed in require block but not used in interface (W-003)
virt.if:  996: (W): Type virt_var_lib_t is listed in require block but not used in interface (W-003)
virt.if: 1246: (W): Definition of declared type virt_bridgehelper_t not found in own module, but in module virt_supplementary (W-011)
virt.if: 1717: (S): Permission macro rw_file_perms does not match class chr_file (S-009)
virt_supplementary.te:283: (S): Permission macro read_file_perms does not match class dir (S-009)

Also, replace spaces in virt_prog_run_bpf interface with tabs and remove
some trailing whitespaces.

Signed-off-by: Vit Mojzis <vmojzis@redhat.com>

5umm3r15 committed:
Update policy for libvirt
